File: control/drone_course_ws/src/exercise_2/src/exercise_2.cpp
// Copyright 2026 Universidad Politécnica de Madrid
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <algorithm>
#include <exercise_2/exercise_2.hpp>

namespace drone_course
{

  DroneCourseExercise2::DroneCourseExercise2(
      const std::string &node_name,
      const rclcpp::NodeOptions &options)
      : rclcpp::Node(node_name, options)
  {
    // QoS profiles
    rclcpp::QoS reliable_qos = rclcpp::QoS(10).reliable();
    rclcpp::QoS best_effort_qos = rclcpp::QoS(10).best_effort();

    callback_group_ =
        this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);

    // Suscribers

    pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
        "/drone0/self_localization/pose", best_effort_qos,
        std::bind(&DroneCourseExercise2::state_subscription_callback, this, std::placeholders::_1));

    // Publishers

    vel_pub_ = this->create_publisher<geometry_msgs::msg::TwistStamped>(
        "/drone0/motion_reference/twist", reliable_qos);

    // Services clients

    control_mode_service_client_ = this->create_client<as2_msgs::srv::SetControlMode>(
        "/drone0/controller/set_control_mode", rmw_qos_profile_services_default, callback_group_);

    path_service_client_ = this->create_client<drone_course_msgs::srv::RequestPath>(
        "/request_path", rmw_qos_profile_services_default, callback_group_);

    // Timers
    double timer_freq = 100.0;
    timer_ = this->create_wall_timer(
        std::chrono::duration<double>(1.0f / timer_freq),
        std::bind(&DroneCourseExercise2::timer_callback, this));

    dt_ = 1.0f / timer_freq;

    RCLCPP_INFO(this->get_logger(), "DroneCourseExercise2 initialized\n");
  }

  DroneCourseExercise2::~DroneCourseExercise2() {}

  void DroneCourseExercise2::timer_callback()
  {
    if (!path_received_)
    {
      // (TODO) Exercise 1: Send the request and process the path message
      path_service_request_ = std::make_shared<drone_course_msgs::srv::RequestPath::Request>();
      rclcpp::Client<drone_course_msgs::srv::RequestPath>::SharedFuture future =
          path_service_client_->async_send_request(path_service_request_).future.share();
      future.wait();
      std::vector<drone_course_msgs::msg::Point> path = future.get()->path;
      if (path.size() > 0)
      {
        RCLCPP_INFO(this->get_logger(), "Path received successfully");
        // Process the received path here
        for (size_t i = 0; i < path.size(); i++)
        {
          path_[3 * i] = path[i].x;
          path_[3 * i + 1] = path[i].y;
          path_[3 * i + 2] = path[i].z;
        }
        path_received_ = true;
      }
      else
      {
        RCLCPP_ERROR(this->get_logger(), "Failed to get path");
        path_received_ = false;
      }
    }

    // Check if control mode is set, if not, call service
    if (!control_mode_set_)
    {
      RCLCPP_INFO(this->get_logger(), "Calling control mode service...");
      auto control_mode_request = std::make_shared<as2_msgs::srv::SetControlMode::Request>();
      control_mode_request->control_mode.control_mode = as2_msgs::msg::ControlMode::SPEED;
      control_mode_request->control_mode.yaw_mode = as2_msgs::msg::ControlMode::YAW_SPEED;
      control_mode_request->control_mode.reference_frame =
          as2_msgs::msg::ControlMode::LOCAL_ENU_FRAME;

      // TODO(Exercise 1): Send control mode service request
      rclcpp::Client<as2_msgs::srv::SetControlMode>::SharedFuture future =
          control_mode_service_client_->async_send_request(
                                          control_mode_request)
              .future.share();
      future.wait();
      as2_msgs::srv::SetControlMode::Response::SharedPtr response = future.get();
      if (response->success)
      {
        RCLCPP_INFO(this->get_logger(), "Control mode set successfully");
        control_mode_set_ = true;
      }
      else
      {
        RCLCPP_ERROR(this->get_logger(), "Failed to set control mode");
        control_mode_set_ = false;
      }
    }

    // Desired position reference
    std::array<double, 3> position_ref = {path_[3 * path_index_],
                                          path_[3 * path_index_ + 1],
                                          path_[3 * path_index_ + 2]};

    // Read current drone state
    double current_x = state_pose_.pose.position.x;
    double current_y = state_pose_.pose.position.y;
    double current_z = state_pose_.pose.position.z;

    double position_error_x = position_ref[0] - current_x;
    double position_error_y = position_ref[1] - current_y;
    double position_error_z = position_ref[2] - current_z;
    double position_error_norm = std::sqrt(
        position_error_x * position_error_x + position_error_y * position_error_y + position_error_z * position_error_z);

    // Compute velocity commands
    double Kp = 3.5;
    double Ki = 0.1;
    double Kd = 1.0;
    const double max_velocity = 4.0;

    const double max_integral = max_velocity / Ki;
    integral_x_ = std::clamp(integral_x_ + position_error_x * dt_, -max_integral, max_integral);
    integral_y_ = std::clamp(integral_y_ + position_error_y * dt_, -max_integral, max_integral);
    integral_z_ = std::clamp(integral_z_ + position_error_z * dt_, -max_integral, max_integral);

    // Skip derivative on first call to avoid spike from uninitialized previous_error
    double derivative_x = 0;
    double derivative_y = 0;
    double derivative_z = 0;
    if (pid_initialized_)
    {
      derivative_x = (position_error_x - previous_error_x_) / dt_;
      derivative_y = (position_error_y - previous_error_y_) / dt_;
      derivative_z = (position_error_z - previous_error_z_) / dt_;
    }
    else
    {
      pid_initialized_ = true;
    }

    previous_error_x_ = position_error_x;
    previous_error_y_ = position_error_y;
    previous_error_z_ = position_error_z;

    double velocity_x = std::clamp(
        Kp * position_error_x + Ki * integral_x_ + Kd * derivative_x,
        -max_velocity, max_velocity);
    double velocity_y = std::clamp(
        Kp * position_error_y + Ki * integral_y_ + Kd * derivative_y,
        -max_velocity, max_velocity);
    double velocity_z = std::clamp(
        Kp * position_error_z + Ki * integral_z_ + Kd * derivative_z,
        -max_velocity, max_velocity);

    // Generate motion reference command
    geometry_msgs::msg::TwistStamped velocity_msg;
    velocity_msg.header.stamp = this->get_clock()->now();
    velocity_msg.header.frame_id = "earth";
    velocity_msg.twist.linear.x = velocity_x;
    velocity_msg.twist.linear.y = velocity_y;
    velocity_msg.twist.linear.z = velocity_z;
    velocity_msg.twist.angular.x = 0.0;
    velocity_msg.twist.angular.y = 0.0;
    velocity_msg.twist.angular.z = 0.0;

    if (vel_pub_)
    {
      vel_pub_->publish(velocity_msg);
    }

    if (position_error_norm < 0.2)
    {
      RCLCPP_INFO(this->get_logger(), "Drone has reach the target position");
      path_index_ = (path_index_ + 1) % (path_.size() / 3);
      integral_x_ = 0.0;
      integral_y_ = 0.0;
      integral_z_ = 0.0;
    }
  }

  void DroneCourseExercise2::state_subscription_callback(
      const geometry_msgs::msg::PoseStamped::SharedPtr msg)
  {
    state_pose_ = *msg;
  }
} // namespace drone_course

File: vision/python_interface/exercise_2.py
# Copyright 2026 Universidad Politécnica de Madrid
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#
#    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
#      contributors may be used to endorse or promote products derived from
#      this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.


import argparse
import os

import cv2
import numpy as np

CornerLabel = tuple[float, float]


def sort_corners(points):
    return sorted(points, key=lambda p: (p[1], p[0]))

def order_points(pts: np.ndarray) -> np.ndarray:
    xSorted = pts[np.argsort(pts[:, 0]), :]
    
    leftMost = xSorted[:2, :]
    rightMost = xSorted[2:, :]

    leftMost = leftMost[np.argsort(leftMost[:, 1]), :]
    (tl, bl) = leftMost

    rightMost = rightMost[np.argsort(rightMost[:, 1]), :]
    (tr, br) = rightMost

    return np.array([tl, tr, bl, br])

def detect_corners(image: np.ndarray) -> list[CornerLabel]:
    # TODO (Exercise 2): Implement a corner detection algorithm that takes an image as input and returns a list of CornerLabel objects.
    h, w = image.shape[:2]
    hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
    _, s, _ = cv2.split(hsv)
    s = cv2.medianBlur(s,7)

    _,binary_img = cv2.threshold(s,50,255,cv2.THRESH_BINARY)

    kernel = np.ones((15,15))
    binary_img = cv2.morphologyEx(binary_img,cv2.MORPH_CLOSE,kernel)
    contours, _ = cv2.findContours(binary_img, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
    
    sorted_contours = sorted(contours, key=cv2.contourArea, reverse=True)
    
    if len(sorted_contours) < 2:
        return [] 
        
    outer_contour = sorted_contours[0]
    inner_contour = sorted_contours[1]
    
    corners = [None] * 8
    
    for idx, contour in enumerate([outer_contour, inner_contour]):
        peri = cv2.arcLength(contour, True)
        approx = cv2.approxPolyDP(contour, 0.05 * peri, True)
        
        if len(approx) == 4:
            pts = approx.reshape(4, 2)
            tl, tr, bl, br = order_points(pts)

            if idx == 0:
                corners[0] = (float(tl[0]) / w, float(tl[1]) / h)
                corners[1] = (float(tr[0]) / w, float(tr[1]) / h)
                corners[6] = (float(bl[0]) / w, float(bl[1]) / h)
                corners[7] = (float(br[0]) / w, float(br[1]) / h)
            else:
                corners[2] = (float(tl[0]) / w, float(tl[1]) / h)
                corners[3] = (float(tr[0]) / w, float(tr[1]) / h)
                corners[4] = (float(bl[0]) / w, float(bl[1]) / h)
                corners[5] = (float(br[0]) / w, float(br[1]) / h)
                
    if all(c is not None for c in corners):
        return corners
    return []


def load_dataset(dataset_path: str) -> list[tuple[cv2.Mat, list[CornerLabel]]]:
    dataset = []
    images_path = os.path.join(dataset_path, 'images')
    if not os.path.exists(images_path):
        print(f'Images directory not found: {images_path}')
        return dataset

    labels_path = os.path.join(dataset_path, 'labels')
    if not os.path.exists(labels_path):
        print(f'Labels directory not found: {labels_path}')
        return dataset

    for filename in os.listdir(images_path):
        if filename.endswith('.jpg') or filename.endswith('.png'):
            image_path = os.path.join(images_path, filename)
            image: np.ndarray | None = cv2.imread(image_path)
            if image is None:
                print(f'Failed to load image: {image_path}')
                continue
            h, w = image.shape[:2]

            label_path = os.path.join(
                labels_path, filename.replace('.jpg', '.txt').replace('.png', '.txt')
            )
            labels: list[CornerLabel] = []
            with open(label_path, 'r') as f:
                for line in f:
                    parts = line.strip().split()
                    parts = parts[5:]
                    for i in range(0, len(parts), 3):
                        x = float(parts[i])
                        y = float(parts[i + 1])
                        labels.append((x, y))

            labels = sort_corners(labels)

            dataset.append((image, labels))
    return dataset


def parse_arguments():
    parser = argparse.ArgumentParser(
        description='Load a dataset of images and segmentation labels.'
    )
    parser.add_argument(
        '--dataset_path',
        type=str,
        help='Path to the dataset directory',
        default='../datasets/yolo_pose',
    )
    return parser.parse_args()


if __name__ == '__main__':
    args = parse_arguments()
    dataset = load_dataset(args.dataset_path)
    print(f'Loaded {len(dataset)} images and their corresponding labels.')

    result: list[list[CornerLabel]] = []
    for image, labels in dataset:
        corners = detect_corners(image)
        corners = sort_corners(corners)
        result.append(corners)

    distances = [1]
    for corners, labels in zip(result, dataset):
        for corner, label in zip(corners, labels[1]):
            distance = np.sqrt((corner[0] - label[0]) ** 2 + (corner[1] - label[1]) ** 2)
            distances.append(distance)

    print(f'Average distance: {np.mean(distances)}')
